src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java

```diff
@@ -0,0 +1,356 @@
+            TokenTypes.RECORD_DEF,
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        // Instantiate new component counter
+        final ComponentCounter counter = new ComponentCounter();
+
+        // Count record components in this record definition
+        countComponents(counter, ast);
+
+        // Check counter and report violations
+        checkCounter(counter, ast);
+    }
+
+    /**
+     * Determine the visibility modifier, then iterate through the record component list
+     * and increment the counter.
+     *
+     * @param counter the ComponentCounter instance for this check
+     * @param ast the ast to check
+     *
+     */
+    private static void countComponents(ComponentCounter counter, DetailAST ast) {
+        final DetailAST recordComponents =
+                ast.findFirstToken(TokenTypes.RECORD_COMPONENTS);
+        final DetailAST modifiers =
+                ast.findFirstToken(TokenTypes.MODIFIERS);
+
+        counter.setScope(ScopeUtil.getScopeFromMods(modifiers));
+
+        DetailAST componentToCount =
+                recordComponents.findFirstToken(TokenTypes.RECORD_COMPONENT_DEF);
+
+        while (componentToCount != null) {
+
+            if (componentToCount.getType() == TokenTypes.RECORD_COMPONENT_DEF) {
+                counter.increment();
+            }
+            componentToCount = componentToCount.getNextSibling();
+
+        }
+
+    }
+
+    /**
+     * Check the counter and report violations.
+     *
+     * @param counter the component counter to check
+     * @param ast to report violations against.
+     */
+    private void checkCounter(ComponentCounter counter, DetailAST ast) {
+        final Scope scope = counter.getScope();
+        if (scope == Scope.PRIVATE) {
+            checkMax(maxPrivate, counter.getTotal(),
+                    MSG_PRIVATE_COMPONENTS, ast);
+        }
+        else if (scope == Scope.PUBLIC) {
+            checkMax(maxPublic, counter.getTotal(),
+                    MSG_PUBLIC_COMPONENTS, ast);
+        }
+        else if (scope == Scope.PROTECTED) {
+            checkMax(maxProtected, counter.getTotal(),
+                    MSG_PROTECTED_COMPONENTS, ast);
+        }
+        else {
+            checkMax(maxPackage, counter.getTotal(),
+                    MSG_PACKAGE_COMPONENTS, ast);
+        }
+    }
+
+    /**
+     * Method for reporting if a maximum has been exceeded.
+     *
+     * @param max the maximum allowed value
+     * @param value the actual value
+     * @param msg the message to log. Takes two arguments of value and maximum.
+     * @param ast the AST to associate with the message.
+     */
+    private void checkMax(int max, int value, String msg, DetailAST ast) {
+        if (max < value) {
+            log(ast, msg, value, max);
+        }
+    }
+
+    /**
+     * Counter class used to track the scope of the record definition and the number of
+     * components in the record component list.
+     *
```

Comment:
Extra blank line in the end of the comment

Reply:
Fixed.